File: src/core/command-registry.ts
/**
 * Command Registry
 *
 * Utilities for loading, validating, and managing custom commands.
 * Commands are markdown files in commands/src/ that define agent instructions.
 */

import * as fs from 'fs';
import * as path from 'path';
import Ajv, { ErrorObject } from 'ajv';
import commandSchema from '../../schemas/command-definition.schema.json';
import { getAllAgents, getAgentById, type AgentDefinition } from './agent-registry';

/** Default commands source directory relative to project root */
const DEFAULT_COMMANDS_DIR = 'commands/src';

/** AJV instance for schema validation */
const ajv = new Ajv({ allErrors: true });
const validateCommandSchemaFn = ajv.compile<CommandDefinition>(commandSchema as Record<string, unknown>);

/**
 * Command definition interface matching the JSON schema
 */
export interface CommandDefinition {
  id: string;
  name: string;
  description: string;
  scriptPath: string;
  agentCompatibility: string[];
  requiresGitHub: boolean;
  outReferences: string[];
  category: 'workflow' | 'git' | 'documentation' | 'utility';
  template?: string;
  characterCount: number;
  wordCount: number;
  sourcePath: string;
}

/**
 * Validation result for command operations
 */
export interface CommandValidationResult {
  valid: boolean;
  errors: CommandValidationError[];
  warnings: CommandValidationError[];
}

/**
 * Individual validation error
 */
export interface CommandValidationError {
  commandId: string;
  message: string;
  severity: 'error' | 'warning';
  file?: string;
}

/**
 * Format AJV validation error for display
 */
function formatSchemaError(err: ErrorObject): string {
  const location = err.instancePath || '/';
  const message = err.message ?? 'validation error';
  return `${location} ${message}`.trim();
}

/**
 * Get the commands directory path
 */
export function getCommandsDirectory(baseDir?: string): string {
  const base = baseDir || process.env.AGENTSMD_HOME || process.cwd();
  return path.join(base, DEFAULT_COMMANDS_DIR);
}

/**
 * Count words in text
 */
function countWords(text: string): number {
  return text.split(/\s+/).filter(word => word.length > 0).length;
}

/**
 * Convert filename to command ID (kebab-case)
 */
function filenameToId(filename: string): string {
  return path.basename(filename, '.md');
}

/**
 * Convert command ID to display name (Title Case)
 */
function idToName(id: string): string {
  return id
    .split('-')
    .map(word => word.charAt(0).toUpperCase() + word.slice(1))
    .join(' ');
}

/**
 * Extract script path from command markdown content
 * Looks for patterns like: python3 ~/.agentsmd/scripts/script_name.py
 */
function extractScriptPath(content: string): string {
  const patterns = [
    /python3?\s+~\/\.agentsmd\/scripts\/(\w+\.py)/,
    /Run:\s*`python3?\s+~\/\.agentsmd\/scripts\/(\w+\.py)`/,
    /`python3?\s+~\/\.agentsmd\/scripts\/(\w+\.py)`/
  ];

  for (const pattern of patterns) {
    const match = content.match(pattern);
    if (match) {
      return `~/.agentsmd/scripts/${match[1]}`;
    }
  }

  return '';
}

/**
 * Extract out-references from command markdown content
 * Finds markdown links to rule-packs/, docs/, templates/, etc.
 */
function extractOutReferences(content: string): string[] {
  const references: string[] = [];
  
  // Match markdown links: [text](path)
  const linkPattern = /\[([^\]]+)\]\(([^)]+)\)/g;
  let match;

  while ((match = linkPattern.exec(content)) !== null) {
    const linkPath = match[2];
    // Filter for relevant paths (rule-packs, docs, templates)
    if (
      linkPath.includes('rule-packs/') ||
      linkPath.includes('docs/') ||
      linkPath.includes('templates/')
    ) {
      // Normalize path (remove relative prefixes)
      const normalized = linkPath.replace(/^\.\.\/+/, '').replace(/^\.\//, '');
      if (!references.includes(normalized)) {
        references.push(normalized);
      }
    }
  }

  return references;
}

/**
 * Determine command category based on content and purpose
 */
function determineCategory(id: string, content: string): 'workflow' | 'git' | 'documentation' | 'utility' {
  const lowerContent = content.toLowerCase();
  const lowerId = id.toLowerCase();

  // Workflow commands (issue, pr, branch creation)
  if (
    lowerId.includes('issue') ||
    lowerId.includes('pr') ||
    lowerId.includes('branch') ||
    lowerContent.includes('github issue') ||
    lowerContent.includes('pull request')
  ) {
    return 'workflow';
  }

  // Git commands (status, push, check operations)
  if (
    lowerId.includes('status') ||
    lowerId.includes('push') ||
    lowerId.includes('check') ||
    lowerId.includes('protect') ||
    lowerContent.includes('git status') ||
    lowerContent.includes('workflow status')
  ) {
    return 'git';
  }

  // Documentation commands (walkthrough, docs)
  if (
    lowerId.includes('walkthrough') ||
    lowerId.includes('doc') ||
    lowerContent.includes('documentation') ||
    lowerContent.includes('walkthrough document')
  ) {
    return 'documentation';
  }

  // Default to utility
  return 'utility';
}

/**
 * Check if command requires GitHub authentication
 */
function requiresGitHub(content: string): boolean {
  const indicators = [
    'gh issue',
    'gh pr',
    'GitHub issue',
    'GitHub CLI',
    'github.com',
    'check_auth.py',
    'Creates GitHub issue'
  ];

  return indicators.some(indicator => content.includes(indicator));
}

/**
 * Determine agent compatibility based on command characteristics
 * Returns array of agent IDs, empty array means all agents
 */
function determineAgentCompatibility(content: string, category: string): string[] {
  // All agents support workflow commands by default
  // CLI-specific commands might only work with certain agents
  
  const agents = getAllAgents();
  const compatible: string[] = [];

  // Check for CLI-only patterns
  const isCLIOnly = content.includes('--cli') || content.includes('command line only');
  
  if (isCLIOnly) {
    // Only agents with CLI command format
    for (const agent of agents) {
      if (agent.commandFormat === 'cli') {
        compatible.push(agent.id);
      }
    }
    return compatible;
  }

  // All agents are compatible by default
  return [];
}

/**
 * Extract template section from command content (if present)
 * Templates are typically in fenced code blocks or after "Template:" headings
 */
function extractTemplate(content: string): string | undefined {
  // Look for template section
  const templateMatch = content.match(/\*\*(?:Walkthrough\s+)?Template:\*\*\s*([\s\S]*?)(?=\n\*\*|\n##|$)/i);
  if (templateMatch) {
    return templateMatch[1].trim();
  }

  // Look for template in code block
  const codeBlockMatch = content.match(/```(?:markdown)?\s*\n([\s\S]*?)```/);
  if (codeBlockMatch && codeBlockMatch[1].includes('##')) {
    return codeBlockMatch[1].trim();
  }

  return undefined;
}

/**
 * Load a command definition from a markdown file
 */
export function loadCommandFromMarkdown(filePath: string): CommandDefinition {
  if (!fs.existsSync(filePath)) {
    throw new Error(`Command file not found: ${filePath}`);
  }

  const content = fs.readFileSync(filePath, 'utf-8');
  const id = filenameToId(filePath);
  const name = idToName(id);
  
  // First line is the description
  const lines = content.split('\n');
  const description = lines[0].trim();

  const scriptPath = extractScriptPath(content);
  const outReferences = extractOutReferences(content);
  const category = determineCategory(id, content);
  const template = extractTemplate(content);
  const agentCompatibility = determineAgentCompatibility(content, category);

  const command: CommandDefinition = {
    id,
    name,
    description,
    scriptPath,
    agentCompatibility,
    requiresGitHub: requiresGitHub(content),
    outReferences,
    category,
    characterCount: content.length,
    wordCount: countWords(content),
    sourcePath: filePath
  };

  if (template) {
    command.template = template;
  }

  return command;
}

/**
 * Get all commands from the commands directory
 */
export function getAllCommands(baseDir?: string): CommandDefinition[] {
  const commandsDir = getCommandsDirectory(baseDir);

  if (!fs.existsSync(commandsDir)) {
    return [];
  }

  const commands: CommandDefinition[] = [];
  const entries = fs.readdirSync(commandsDir, { withFileTypes: true });

  for (const entry of entries) {
    if (entry.isFile() && entry.name.endsWith('.md')) {
      const filePath = path.join(commandsDir, entry.name);
      try {
        const command = loadCommandFromMarkdown(filePath);
        commands.push(command);
      } catch (error) {
        // Skip invalid command files
        console.warn(`Skipping invalid command file: ${entry.name}`, error);
      }
    }
  }

  return commands;
}

/**
 * Get a command by its ID
 */
export function getCommandById(id: string, baseDir?: string): CommandDefinition | undefined {
  const commandsDir = getCommandsDirectory(baseDir);
  const filePath = path.join(commandsDir, `${id}.md`);

  if (!fs.existsSync(filePath)) {
    return undefined;
  }

  try {
    return loadCommandFromMarkdown(filePath);
  } catch {
    return undefined;
  }
}

/**
 * Get commands by category
 */
export function getCommandsByCategory(
  category: 'workflow' | 'git' | 'documentation' | 'utility',
  baseDir?: string
): CommandDefinition[] {
  return getAllCommands(baseDir).filter(cmd => cmd.category === category);
}

/**
 * Get commands compatible with a specific agent
 * If command.agentCompatibility is empty, it's compatible with all agents
 */
export function getCommandsForAgent(agentId: string, baseDir?: string): CommandDefinition[] {
  const agent = getAgentById(agentId);
  if (!agent) {
    return [];
  }

  return getAllCommands(baseDir).filter(cmd => {
    // Empty agentCompatibility means all agents
    if (cmd.agentCompatibility.length === 0) {
      return true;
    }
    return cmd.agentCompatibility.includes(agentId);
  });
}

/**
 * Read raw command content
 */
export function getCommandContent(id: string, baseDir?: string): string | undefined {
  const commandsDir = getCommandsDirectory(baseDir);
  const filePath = path.join(commandsDir, `${id}.md`);

  if (!fs.existsSync(filePath)) {
    return undefined;
  }

  return fs.readFileSync(filePath, 'utf-8');
}

/**
 * Validate a command definition against the schema
 */
export function validateCommand(command: CommandDefinition): CommandValidationResult {
  const errors: CommandValidationError[] = [];
  const warnings: CommandValidationError[] = [];

  // Schema validation
  const isValid = validateCommandSchemaFn(command);
  if (!isValid && validateCommandSchemaFn.errors) {
    for (const err of validateCommandSchemaFn.errors) {
      errors.push({
        commandId: command.id,
        message: `Schema violation: ${formatSchemaError(err)}`,
        severity: 'error'
      });
    }
  }

  // Check if script path is set
  if (!command.scriptPath) {
    warnings.push({
      commandId: command.id,
      message: 'No script path detected in command content',
      severity: 'warning'
    });
  }

  // Check out-references exist
  for (const ref of command.outReferences) {
    const baseDir = path.dirname(command.sourcePath);
    // Try to resolve relative path from command source
    let refPath = ref;
    if (ref.startsWith('rule-packs/') || ref.startsWith('docs/')) {
      // These are relative to project root
      refPath = path.join(process.cwd(), ref);
    }
    
    if (!fs.existsSync(refPath) && !fs.existsSync(path.join(baseDir, '..', '..', ref))) {
      warnings.push({
        commandId: command.id,
        message: `Out-reference not found: ${ref}`,
        severity: 'warning',
        file: ref
      });
    }
  }

  return {
    valid: errors.length === 0,
    errors,
    warnings
  };
}

/**
 * Validate command compatibility with a specific agent
 */
export function validateCommandForAgent(
  commandId: string,
  agentId: string,
  baseDir?: string
): { compatible: boolean; reason?: string } {
  const command = getCommandById(commandId, baseDir);
  if (!command) {
    return { compatible: false, reason: `Command not found: ${commandId}` };
  }

  const agent = getAgentById(agentId);
  if (!agent) {
    return { compatible: false, reason: `Agent not found: ${agentId}` };
  }

  // Check explicit compatibility list
  if (command.agentCompatibility.length > 0 && !command.agentCompatibility.includes(agentId)) {
    return {
      compatible: false,
      reason: `Command ${commandId} is not compatible with agent ${agentId}`
    };
  }

  // Check if agent supports out-references when command has them
  if (command.outReferences.length > 0 && !agent.characterLimits.supportsOutReferences) {
    return {
      compatible: false,
      reason: `Agent ${agentId} does not support out-references required by ${commandId}`
    };
  }

  // Check character limits
  if (agent.characterLimits.maxChars !== null && command.characterCount > agent.characterLimits.maxChars) {
    return {
      compatible: false,
      reason: `Command ${commandId} exceeds character limit for ${agentId} (${command.characterCount} > ${agent.characterLimits.maxChars})`
    };
  }

  return { compatible: true };
}

/**
 * Export commands as JSON (for Rust/Python consumption)
 */
export function exportForPython(): string {
  const commands = getAllCommands();
  return JSON.stringify(commands, null, 2);
}

/**
 * Check if a command exists
 */
export function commandExists(id: string, baseDir?: string): boolean {
  const commandsDir = getCommandsDirectory(baseDir);
  const filePath = path.join(commandsDir, `${id}.md`);
  return fs.existsSync(filePath);
}

// CLI entry point
if (require.main === module) {
  try {
    const shouldExport = process.argv.includes('--export-json');
    const listOnly = process.argv.includes('--list');
    
    if (shouldExport) {
      process.stdout.write(exportForPython());
    } else if (listOnly) {
      const commands = getAllCommands();
      for (const cmd of commands) {
        console.log(`${cmd.id}: ${cmd.description} [${cmd.category}]`);
      }
    } else {
      console.log(JSON.stringify(getAllCommands(), null, 2));
    }
  } catch (err) {
    console.error((err as Error).message);
    process.exit(1);
  }
}

File: desktop/src-tauri/src/command_registry.rs
use crate::fs_manager;
use crate::types::*;
use once_cell::sync::Lazy;
use regex::Regex;
use std::collections::HashMap;
use std::fs;
use std::path::PathBuf;
use std::sync::Mutex;
use serde_json;

/// Cached commands to avoid re-parsing on every request
static COMMAND_CACHE: Lazy<Mutex<Option<Vec<CommandMetadata>>>> = Lazy::new(|| Mutex::new(None));

/// Get the commands source directory path
pub fn get_commands_directory() -> PathBuf {
    fs_manager::get_agentsmd_home().join("commands").join("src")
}

/// Count words in text
fn count_words(text: &str) -> u64 {
    text.split_whitespace().count() as u64
}

/// Convert filename to command ID (kebab-case)
fn filename_to_id(filename: &str) -> String {
    filename.trim_end_matches(".md").to_string()
}

/// Convert command ID to display name (Title Case)
fn id_to_name(id: &str) -> String {
    id.split('-')
        .map(|word| {
            let mut chars = word.chars();
            match chars.next() {
                None => String::new(),
                Some(first) => first.to_uppercase().chain(chars).collect(),
            }
        })
        .collect::<Vec<_>>()
        .join(" ")
}

/// Extract script path from command markdown content
fn extract_script_path(content: &str) -> String {
    static PATTERNS: Lazy<Vec<Regex>> = Lazy::new(|| {
        vec![
            Regex::new(r"python3?\s+~/\.agentsmd/scripts/(\w+\.py)").unwrap(),
            Regex::new(r"Run:\s*`python3?\s+~/\.agentsmd/scripts/(\w+\.py)`").unwrap(),
            Regex::new(r"`python3?\s+~/\.agentsmd/scripts/(\w+\.py)`").unwrap(),
        ]
    });

    for pattern in PATTERNS.iter() {
        if let Some(caps) = pattern.captures(content) {
            if let Some(script_name) = caps.get(1) {
                return format!("~/.agentsmd/scripts/{}", script_name.as_str());
            }
        }
    }

    String::new()
}

/// Extract out-references from command markdown content
fn extract_out_references(content: &str) -> Vec<String> {
    static LINK_PATTERN: Lazy<Regex> =
        Lazy::new(|| Regex::new(r"\[([^\]]+)\]\(([^)]+)\)").unwrap());

    let mut references = Vec::new();

    for caps in LINK_PATTERN.captures_iter(content) {
        if let Some(link_path) = caps.get(2) {
            let path = link_path.as_str();
            // Filter for relevant paths
            if path.contains("rule-packs/") || path.contains("docs/") || path.contains("templates/")
            {
                // Normalize path (remove relative prefixes)
                let normalized = path
                    .trim_start_matches("../")
                    .trim_start_matches("./")
                    .to_string();
                if !references.contains(&normalized) {
                    references.push(normalized);
                }
            }
        }
    }

    references
}

/// Path to persisted out-reference overrides for commands
fn command_out_ref_overrides_path() -> PathBuf {
    fs_manager::get_agentsmd_home()
        .join("commands")
        .join("out-references.json")
}

/// Load persisted out-reference overrides for commands
fn load_command_out_ref_overrides() -> HashMap<String, Vec<String>> {
    let path = command_out_ref_overrides_path();
    if !path.exists() {
        return HashMap::new();
    }

    match fs::read_to_string(&path) {
        Ok(content) => serde_json::from_str(&content).unwrap_or_default(),
        Err(e) => {
            log::warn!("Failed to read command out-reference overrides: {}", e);
            HashMap::new()
        }
    }
}

/// Persist command out-reference overrides
fn save_command_out_ref_overrides(map: &HashMap<String, Vec<String>>) -> Result<(), String> {
    let path = command_out_ref_overrides_path();
    if let Some(parent) = path.parent() {
        fs::create_dir_all(parent)
            .map_err(|e| format!("Failed to create overrides directory: {}", e))?;
    }

    let content = serde_json::to_string_pretty(map)
        .map_err(|e| format!("Failed to serialize overrides: {}", e))?;
    fs::write(&path, content)
        .map_err(|e| format!("Failed to write overrides file: {}", e))
}

/// Update a command's out-references and persist the override
pub fn update_command_out_references(command_id: &str, references: Vec<String>) -> Result<(), String> {
    let mut overrides = load_command_out_ref_overrides();

    let mut unique_refs = references;
    unique_refs.sort();
    unique_refs.dedup();

    overrides.insert(command_id.to_string(), unique_refs);
    save_command_out_ref_overrides(&overrides)?;

    // Clear cache so subsequent reads return updated metadata
    clear_cache();
    Ok(())
}

/// Determine command category based on content and purpose
fn determine_category(id: &str, content: &str) -> String {
    let lower_content = content.to_lowercase();
    let lower_id = id.to_lowercase();

    // Workflow commands (issue, pr, branch creation)
    if lower_id.contains("issue")
        || lower_id.contains("pr")
        || lower_id.contains("branch")
        || lower_content.contains("github issue")
        || lower_content.contains("pull request")
    {
        return "workflow".to_string();
    }

    // Git commands (status, push, check operations)
    if lower_id.contains("status")
        || lower_id.contains("push")
        || lower_id.contains("check")
        || lower_id.contains("protect")
        || lower_content.contains("git status")
        || lower_content.contains("workflow status")
    {
        return "git".to_string();
    }

    // Documentation commands (walkthrough, docs)
    if lower_id.contains("walkthrough")
        || lower_id.contains("doc")
        || lower_content.contains("documentation")
        || lower_content.contains("walkthrough document")
    {
        return "documentation".to_string();
    }

    // Default to utility
    "utility".to_string()
}

/// Check if command requires GitHub authentication
fn requires_github(content: &str) -> bool {
    let indicators = [
        "gh issue",
        "gh pr",
        "GitHub issue",
        "GitHub CLI",
        "github.com",
        "check_auth.py",
        "Creates GitHub issue",
    ];

    indicators.iter().any(|indicator| content.contains(indicator))
}

/// Extract template section from command content
fn extract_template(content: &str) -> Option<String> {
    static TEMPLATE_PATTERN: Lazy<Regex> = Lazy::new(|| {
        Regex::new(r"(?i)\*\*(?:Walkthrough\s+)?Template:\*\*\s*([\s\S]*?)(?:\n\*\*|\n##|$)").unwrap()
    });

    static CODE_BLOCK_PATTERN: Lazy<Regex> =
        Lazy::new(|| Regex::new(r"```(?:markdown)?\s*\n([\s\S]*?)```").unwrap());

    // Look for template section
    if let Some(caps) = TEMPLATE_PATTERN.captures(content) {
        if let Some(template) = caps.get(1) {
            return Some(template.as_str().trim().to_string());
        }
    }

    // Look for template in code block
    if let Some(caps) = CODE_BLOCK_PATTERN.captures(content) {
        if let Some(code_block) = caps.get(1) {
            let code = code_block.as_str();
            if code.contains("##") {
                return Some(code.trim().to_string());
            }
        }
    }

    None
}

/// Load a command from a markdown file
fn load_command_from_file(file_path: &PathBuf) -> Result<CommandMetadata, String> {
    if !file_path.exists() {
        return Err(format!("Command file not found: {:?}", file_path));
    }

    let content = fs::read_to_string(file_path)
        .map_err(|e| format!("Failed to read command file: {}", e))?;

    let filename = file_path
        .file_name()
        .and_then(|n| n.to_str())
        .unwrap_or("");
    let id = filename_to_id(filename);
    let name = id_to_name(&id);

    // First line is the description
    let description = content
        .lines()
        .next()
        .unwrap_or("")
        .trim()
        .to_string();

    let script_path = extract_script_path(&content);
    let out_references = extract_out_references(&content);
    let category = determine_category(&id, &content);
    let template = extract_template(&content);

    Ok(CommandMetadata {
        id,
        name,
        description,
        script_path,
        agent_compatibility: Vec::new(), // Empty means all agents
        requires_github: requires_github(&content),
        out_references,
        category,
        template,
        character_count: content.len() as u64,
        word_count: count_words(&content),
        source_path: file_path.to_string_lossy().to_string(),
    })
}

/// Load all commands from the commands directory
pub fn load_commands() -> Result<Vec<CommandMetadata>, String> {
    // Check cache first
    if let Ok(cache) = COMMAND_CACHE.lock() {
        if let Some(ref commands) = *cache {
            return Ok(commands.clone());
        }
    }

    let overrides = load_command_out_ref_overrides();

    let commands_dir = get_commands_directory();

    if !commands_dir.exists() {
        return Ok(Vec::new());
    }

    let mut commands = Vec::new();

    let entries =
        fs::read_dir(&commands_dir).map_err(|e| format!("Failed to read commands directory: {}", e))?;

    for entry in entries.flatten() {
        let path = entry.path();
        if path.extension().map(|e| e == "md").unwrap_or(false) {
            match load_command_from_file(&path) {
                Ok(mut cmd) => {
                    if let Some(override_refs) = overrides.get(&cmd.id) {
                        cmd.out_references = override_refs.clone();
                    }
                    commands.push(cmd)
                }
                Err(e) => {
                    log::warn!("Skipping invalid command file {:?}: {}", path, e);
                }
            }
        }
    }

    // Update cache
    if let Ok(mut cache) = COMMAND_CACHE.lock() {
        *cache = Some(commands.clone());
    }

    Ok(commands)
}

/// Get a command by its ID
pub fn get_command_by_id(command_id: &str) -> Result<CommandMetadata, String> {
    let commands = load_commands()?;
    commands
        .into_iter()
        .find(|c| c.id == command_id)
        .ok_or_else(|| format!("Command not found: {}", command_id))
}

/// Get commands compatible with a specific agent
pub fn get_commands_for_agent(agent_id: &str) -> Result<Vec<CommandMetadata>, String> {
    let commands = load_commands()?;
    let agents = fs_manager::load_agent_registry()
        .map_err(|e| format!("Failed to load agent registry: {}", e))?;

    // Find the agent
    let agent = agents
        .iter()
        .find(|a| a.id == agent_id)
        .ok_or_else(|| format!("Agent not found: {}", agent_id))?;

    Ok(commands
        .into_iter()
        .filter(|cmd| {
            // Empty agent_compatibility means all agents
            if cmd.agent_compatibility.is_empty() {
                return true;
            }
            cmd.agent_compatibility.contains(&agent_id.to_string())
        })
        .collect())
}

/// Get commands by category
pub fn get_commands_by_category(category: &str) -> Result<Vec<CommandMetadata>, String> {
    let commands = load_commands()?;
    Ok(commands
        .into_iter()
        .filter(|cmd| cmd.category == category)
        .collect())
}

/// Read raw command content
pub fn get_command_content(command_id: &str) -> Result<String, String> {
    let commands_dir = get_commands_directory();
    let file_path = commands_dir.join(format!("{}.md", command_id));

    if !file_path.exists() {
        return Err(format!("Command not found: {}", command_id));
    }

    fs::read_to_string(&file_path).map_err(|e| format!("Failed to read command file: {}", e))
}

/// Validate command compatibility with a specific agent
pub fn validate_command_for_agent(
    command_id: &str,
    agent_id: &str,
) -> Result<CommandCompatibilityResult, String> {
    let command = get_command_by_id(command_id)?;
    let agents = fs_manager::load_agent_registry()
        .map_err(|e| format!("Failed to load agent registry: {}", e))?;

    let agent = agents
        .iter()
        .find(|a| a.id == agent_id)
        .ok_or_else(|| format!("Agent not found: {}", agent_id))?;

    // Check explicit compatibility list
    if !command.agent_compatibility.is_empty() && !command.agent_compatibility.contains(&agent_id.to_string()) {
        return Ok(CommandCompatibilityResult {
            compatible: false,
            reason: Some(format!(
                "Command {} is not compatible with agent {}",
                command_id, agent_id
            )),
        });
    }

    // Check if agent supports out-references when command has them
    if !command.out_references.is_empty() && !agent.character_limits.supports_out_references {
        return Ok(CommandCompatibilityResult {
            compatible: false,
            reason: Some(format!(
                "Agent {} does not support out-references required by {}",
                agent_id, command_id
            )),
        });
    }

    // Check character limits
    if let Some(max_chars) = agent.character_limits.max_chars {
        if command.character_count > max_chars {
            return Ok(CommandCompatibilityResult {
                compatible: false,
                reason: Some(format!(
                    "Command {} exceeds character limit for {} ({} > {})",
                    command_id, agent_id, command.character_count, max_chars
                )),
            });
        }
    }

    Ok(CommandCompatibilityResult {
        compatible: true,
        reason: None,
    })
}

/// Clear the command cache (useful after file changes)
pub fn clear_cache() {
    if let Ok(mut cache) = COMMAND_CACHE.lock() {
        *cache = None;
    }
}

/// Calculate total character budget for a set of commands
pub fn calculate_command_budget(command_ids: &[String]) -> Result<CommandBudgetInfo, String> {
    let mut total_chars: u64 = 0;
    let mut command_breakdown = Vec::new();

    for command_id in command_ids {
        let command = get_command_by_id(command_id)?;
        total_chars += command.character_count;
        command_breakdown.push(CommandBudgetItem {
            command_id: command.id,
            chars: command.character_count,
            words: command.word_count,
        });
    }

    Ok(CommandBudgetInfo {
        total_chars,
        command_breakdown,
    })
}
